staging/vhost-device-can/src/backend.rs

```diff
@@ -0,0 +1,352 @@
+            VhostUserCanBackend::new(lockable_controller.clone())
+                .map_err(Error::CouldNotCreateBackend)?,
+        ));
+        lockable_controller
+            .write()
+            .unwrap()
+            .open_can_socket()
+            .map_err(Error::FailCreateCanControllerSocket)?;
+
+        let read_handle = CanController::start_read_thread(lockable_controller.clone());
+
+        let mut daemon = VhostUserDaemon::new(
+            String::from("vhost-device-can-backend"),
+            vu_can_backend.clone(),
+            GuestMemoryAtomic::new(GuestMemoryMmap::new()),
+        )
+        .map_err(Error::CouldNotCreateDaemon)?;
+
+        /* Start the read thread -- need to handle it after termination */
+        let vring_workers = daemon.get_epoll_handlers();
+        vu_can_backend
+            .read()
+            .unwrap()
+            .set_vring_worker(&vring_workers[0]);
+
+        daemon.serve(&socket).map_err(Error::ServeFailed)?;
+
+        // Terminate the thread which reads CAN messages from "can_devs"
+        lockable_controller.write().unwrap().exit_read_thread();
+
+        // Wait for read thread to exit
+        match read_handle.join() {
+            Ok(_) => info!("The read thread returned successfully"),
+            Err(e) => warn!("The read thread returned the error: {:?}", e),
+        }
+    }
+}
+
+fn start_backend(config: VuCanConfig) -> Result<()> {
+    let mut handles = HashMap::new();
+    let (senders, receiver) = std::sync::mpsc::channel();
+
+    for (thread_id, (socket, can_devs)) in config
+        .generate_socket_paths()
+        .into_iter()
+        .zip(config.can_devices.iter().cloned())
+        .map(|(a, b)| (a, b.to_string()))
+        .enumerate()
+    {
+        println!(
+            "thread_id: {}, socket: {:?}, can_devs: {:?}",
+            thread_id, socket, can_devs,
+        );
+
+        let name = format!("vhu-can-{}", can_devs);
+        let sender = senders.clone();
+        let handle = thread::Builder::new()
+            .name(name.clone())
+            .spawn(move || {
+                let result =
+                    std::panic::catch_unwind(move || start_backend_server(socket, can_devs));
+
+                // Notify the main thread that we are done.
+                sender.send(thread_id).unwrap();
+
+                result.map_err(|e| Error::ThreadPanic(name, e))?
+            })
+            .unwrap();
+        handles.insert(thread_id, handle);
+    }
+
+    while !handles.is_empty() {
+        let thread_id = receiver.recv().unwrap();
+        handles
+            .remove(&thread_id)
+            .unwrap()
+            .join()
+            .map_err(std::panic::resume_unwind)
+            .unwrap()?;
+    }
+
+    Ok(())
+}
+
+pub(crate) fn can_init() {
+    env_logger::init();
+    if let Err(e) = VuCanConfig::try_from(CanArgs::parse()).and_then(start_backend) {
+        error!("{e}");
+        exit(1);
+    }
+}
```

Comment:
I'm fine to have `start_backend` here or in lib.rs, but I'd move the application args in main.rs.
Take a look at sound for example.

Reply:
I followed sound's example and moved the args and helper functions into `main.rs`